---
title: "MILLIPEDE workflow"
author: "Kaixuan Luo"
# date: YYYY-MM-DD
output: html_document
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```

###  Step 1: find candidate binding sites by scanning for motif matches along the genome

* Yeast genome (UCSC June 2008 sacCer2) FASTA files can be downloaded from [UCSC][genome]

[genome]: http://hgdownload.cse.ucsc.edu/goldenPath/sacCer2/chromosomes/

* Background file is [bg_all.tab][bg]

[bg]: data/bg_all.tab

* TF motif files can be found in [tf_coord_updated_macisaac][motif]

[motif]: ../data/tf_coord_updated_macisaac


### Step 2: extract DNase cuts around the candidate binding sites

### Step 3: bin DNase cuts

### Step 4: predict TF binding using trained MILLIPEDE model parameters

